src/evm/protocol/uniswap_v4/state.rs

```diff
@@ -217,50 +227,92 @@ impl UniswapV4State {
     }
 }
 
-#[allow(unused_variables)] //TODO: remove when implemented
 impl ProtocolSim for UniswapV4State {
+    // Not possible to implement correctly with the current interface because we need to know the
+    // swap direction.
     fn fee(&self) -> f64 {
         todo!()
     }
 
-    fn spot_price(
-        &self,
-        base: &crate::models::Token,
-        quote: &crate::models::Token,
-    ) -> Result<f64, SimulationError> {
-        todo!()
+    fn spot_price(&self, base: &Token, quote: &Token) -> Result<f64, SimulationError> {
+        if base < quote {
+            Ok(sqrt_price_q96_to_f64(self.sqrt_price, base.decimals as u32, quote.decimals as u32))
+        } else {
+            Ok(1.0f64 /
+                sqrt_price_q96_to_f64(
+                    self.sqrt_price,
+                    quote.decimals as u32,
+                    base.decimals as u32,
+                ))
+        }
     }
 
     fn get_amount_out(
         &self,
-        amount_in: num_bigint::BigUint,
-        token_in: &crate::models::Token,
-        token_out: &crate::models::Token,
+        amount_in: BigUint,
+        token_in: &Token,
+        token_out: &Token,
     ) -> Result<GetAmountOutResult, SimulationError> {
-        todo!()
+        let zero_for_one = token_in < token_out;
+        let amount_specified = I256::checked_from_sign_and_abs(
+            Sign::Positive,
+            U256::from_be_slice(&amount_in.to_bytes_be()),
+        )
+        .unwrap();
```

Comment:
can you use expect instead of unwrap here? 👀

Reply:
So proud of you, @dianacarvalho1 ❤️ 🤣

Reply:
I've levelled up in my Rust track!! ➕